glean-core/ffi/glean.h

```diff
@@ -69,6 +69,29 @@ typedef const int32_t *RawIntArray;
 
 typedef const char *const *RawStringArray;
 
+/**
+ * A FFI-compatible representation for the PingUploadTask
+ */
+enum FfiPingUploadTask_Tag {
+  Wait,
+  Upload,
+  Done,
+};
+typedef uint8_t FfiPingUploadTask_Tag;
+
+typedef struct {
+  FfiPingUploadTask_Tag tag;
+  const char *uuid;
+  const char *path;
+  const char *body;
+  const char *headers;
```

Comment:
Should this rather be a string array, given the plural?

Reply:
I decided to go with JSON encoded string here because I thought it made more sense. This is what this string will look like in this case (I added line breaks to make it easier to visualize, it won't really have them): 
```json
{
    "X-Client-Type":"Glean",
    "Content-Type":"application/json; charset=utf-8","X-Client-Version":"25.0.0",
    "Date":"2020-02-26 09:54:40.648697 UTC"
}
```

Reply:
I'm not sure this is the best option here. We don't require it to be JSON for sending headers, and Android would require to parse it/break it into a key value map even if it's JSON (see [here](https://searchfox.org/glean/rev/7f4d8b01a47750b2cb633b69f724128c2396b628/glean-core/android/src/main/java/mozilla/telemetry/glean/net/BaseUploader.kt#117) and [here](https://searchfox.org/mozilla-mobile/source/android-components/components/service/glean/src/main/java/mozilla/components/service/glean/net/ConceptFetchHttpUploader.kt#67)).

Since we don't really require JSON, let's report back two arrays: one for the keys and ones for the values. If it makes it more convenient to just keep it JSON to avoid the nasty two arrays passing, make sure to encapsulate and abstract this away to the other FFI side. Please also rename this to something more meaningful such as `headersAsJSON` or `jsonHeaders`. Let's also add a big comment about why we're doing this and not going the key-value map route.

Reply:
I think encoding it as JSON is preferably to the 2-array solution.
That will also make it easier to free it properly (there's already `glean_str_free`).
I do agree we can do the map-to-JSON serialization in glean-ffi instead of doing that in glean-core.

Reply:
> I do agree we can do the map-to-JSON serialization in glean-ffi instead of doing that in glean-core.

The only thing that made me put it in glean-core at first was because to parse it as json on the ffi we would have to add serde_json as a dependency there. Probably not a big deal, but let me know if you think I should go ahead with it.

Reply:
Not a big deal at all, it's a dependency anyway and as it's all compiled together there's no duplication.